p4-16/spec/P4-16-spec.mdk

```diff
@@ -2636,15 +2636,23 @@ These declarations describe two objects:
 ### Extern types { #sec_extern }
 
 []{tex-cmd: "\indent"}
-P4 supports extern object declarations and extern function declarations using the following syntax.
+P4 supports extern object declarations, extern function declarations, and forward extern declarations using the following syntax.
 
 ~ Begin P4Grammar
 externDeclaration
     : optAnnotations EXTERN nonTypeName optTypeParameters '{' methodPrototypes '}'
     | optAnnotations EXTERN functionPrototype ';'
+    | optAnnotations EXTERN name ';'
     ;
 ~ End P4Grammar
 
+The last case is a forward extern declaration, intended for declaring
```

Comment:
I guess this allows mutually-recursive declarations (instantiations). Otherwise this would be impossible.

Reply:
I do not know of an example, other than Chris Dodd's mention during the LDWG meeting that perhaps this was added because of the way that the TNA architecture defines the Register and RegisterAction externs.  He did not sound confident that this was the motivating reason, so it might be some other reason.  I do not see any occurrences of forward declarations of externs in this public file, for example: https://github.com/barefootnetworks/Open-Tofino/blob/master/share/p4c/p4include/tofino.p4

Reply:
Following discussion at the 2021-May-03 LDWG meeting, I have removed all changes related to forward extern declarations from this PR.